zict/lru.py

```diff
@@ -55,19 +55,24 @@ def __setitem__(self, key, value):
 
         weight = self.weight(key, value)
 
-        if weight <= self.n:
+        def set_():
             self.d[key] = value
             self.i += 1
             self.heap[key] = self.i
-
             self.weights[key] = weight
             self.total_weight += weight
-        else:
-            for cb in self.on_evict:
-                cb(key, value)
+            while self.total_weight > self.n:
+                self.evict()
```

Comment:
There are 2 use cases here, when weight > self.n:

1. the sum of all *other* keys in self.d is enough to fill slow. so you will raise on line 72 and again on again on line 65 at some point in the middle of the while iteration. This is desirable.
2. the sum of all *other* keys fits in slow, but the while won't stop until it tries and - we know in advance - fail to store key too into slow. This can be avoided with the following:

```suggestion
            while self.total_weight > self.n and len(self.d) > 1:
                self.evict()
```